### PR TITLE
Feature/buyer protection

### PR DESCRIPTION
**Pull request by prempratapsingh**
## Description
Buyer protection widget is now added to the library with,

1. Figma based UI/UX specs
2. Integration to backend API for fetching and displaying shop specific buyer protection details
3. Unit tests

Also refactored Trustbadge view and widget views for improved unit test coverage. Sonar cloud reports now show `91% code coverage` 🚀 

Xcode unit test coverage reports however shows 92.3% code coverage. I will see why there is a difference b/w XCode code coverage and Sonar cloud code coverage.

### CODE REVIEW

File: Sources/Trustylib/Utilities/GradeCalculator.swift
Comment (superus8r) [inline Sources/Trustylib/Utilities/GradeCalculator.swift:1]: suggestion: this business logic could be part of domain layer or a model for readability.
This can avoid a large number of Util files in future by keeping the business logic of relevant parts together. 🙂

File: Sources/Trustylib/Models/CurrencyCode.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    case euro, gbp
+    
+    // MARK: Public properties
+    
+    /**
+     Returns currency code
+     */
+    var code: String {
+        switch self {
+        case .euro: return "EURO"
```
Comment (superus8r) [inline Sources/Trustylib/Models/CurrencyCode.swift:43]: praise: it's amazing to see how swift and kotlin code are similar, it makes code bases more similar and consistent 👍

**Review by superus8r — CHANGES_REQUESTED** [review]
The code itself looks good 🙂 there is only one thing: It seems like there is code coverage reported for the "Test files". 

chore: Please exclude the test files from calculated coverage report.

Everything else looks good, added some suggestions in the comments. 🙂

**Comment by prempratapsingh** [comment]
> The code itself looks good 🙂 there is only one thing: It seems like there is code coverage reported for the "Test files".
> 
> chore: Please exclude the test files from calculated coverage report.
> 
> Everything else looks good, added some suggestions in the comments. 🙂

hey @superus8r what do you mean by `Test Files`?

XCTest test results and code coverage are based on the actual Swift/Objective-C code which are executed unit tests. It doesn't consider artifect files like mock JSON/Text, etc.

The increase in code coverage % report is because I have refactored the SwiftUI views, Models and ViewModels in a way that more code blocks could be covered during tests runs. I also added good amount of new tests 🙂

**Comment by superus8r** [comment]
Hey @prempratapsingh thank you for your answer.

> what do you mean by Test Files? 

Test Files are files under `Tests` group.

> XCTest test results and code coverage are based on the actual Swift/Objective-C code which are executed unit tests.

Could you please tell which groups (e.g. `Sources`, `Tests`, `Example`, etc) have active test coverage? 🙂

**Comment by prempratapsingh** [comment]
> Hey @prempratapsingh thank you for your answer.
> 
> > what do you mean by Test Files?
> 
> Test Files are files under `Tests` group.
> 
> > XCTest test results and code coverage are based on the actual Swift/Objective-C code which are executed unit tests.
> 
> Could you please tell which groups (e.g. `Sources`, `Tests`, `Example`, etc) have active test coverage? 🙂

@superus8r XCTest doesn't work on the concept of covering a group `Sources`, `Tests`, `Example`, etc.
Rather it uses a TestTarget named `TrustylibTests` where unit test classes (BuyerProtectionDetailsModelTests, BuyerProtectionViewModelTests, etc) are added as pure Swift code. 

So in essence `TrustylibTests` has only Swift test cases (shown in screenshot below) added to it and thus the test reports, code coverage reports are based on these test cases only.

Screenshot 2023-04-03 at 2 15 32 PM

**Comment by superus8r** [comment]
@prempratapsingh 

> XCTest doesn't work on the concept of covering a group Sources, Tests, Example, etc.
Rather it uses a TestTarget named TrustylibTests where unit test classes (BuyerProtectionDetailsModelTests, BuyerProtectionViewModelTests, etc) are added as pure Swift code.

Thank you for clarifying that the `XCTest` uses `TrustylibTests` as TestTarget where unit test classes are added as pure Swift code, I understand that. 🙂✅

> So in essence TrustylibTests has only Swift test cases (shown in screenshot below) added to it and thus the test reports, code coverage reports are based on these test cases only.

I also understand that the `TrustylibTests` only has Swift test cases and you agree that coverage reports must be based on the test cases. ✅
It's great to know you also agree with that, 🙂 however, along with that, the tests are also reporting coverage on themselves. This is extra and needs to be excluded. 

In a nutshell: tests should not report coverage on themselves and this should be excluded 🙂 
I also added a screenshot:
Screenshot 2023-04-03 at 11 01 48

**Comment by superus8r** [comment]
@prempratapsingh 
I approved this for now.
If it's okay, I can fix it myself later. 🙂

**Review by superus8r — APPROVED** [review]

**Comment by prempratapsingh** [comment]
> @prempratapsingh
> 
> > XCTest doesn't work on the concept of covering a group Sources, Tests, Example, etc.
> > Rather it uses a TestTarget named TrustylibTests where unit test classes (BuyerProtectionDetailsModelTests, BuyerProtectionViewModelTests, etc) are added as pure Swift code.
> 
> Thank you for clarifying that the `XCTest` uses `TrustylibTests` as TestTarget where unit test classes are added as pure Swift code, I understand that. 🙂✅
> 
> > So in essence TrustylibTests has only Swift test cases (shown in screenshot below) added to it and thus the test reports, code coverage reports are based on these test cases only.
> 
> I also understand that the `TrustylibTests` only has Swift test cases and you agree that coverage reports must be based on the test cases. ✅ It's great to know you also agree with that, 🙂 however, along with that, the tests are also reporting coverage on themselves. This is extra and needs to be excluded.
> 
> In a nutshell: tests should not report coverage on themselves and this should be excluded 🙂 @irockel could we please get a third opinion on this? I also added a screenshot: Screenshot 2023-04-03 at 11 01 48

@superus8r the CircleCI configuration for groups to be covered (that you showed in attached screenshot) hasn't been changed. It is the same as when the iOS trustylib code coverage was 72%

The increase in code coverage reports is primarily due to the refactoring I did last week for Views, Models, ViewModel and other related layers. Please see this code commit for reference - https://github.com/trustedshops-public/etrusted-ios-trustbadge-library/commit/3b04f6dcbd6db35ca580913cd815cd15a9d0cd96

**Comment by prempratapsingh** [comment]
> @prempratapsingh I approved this for now. If it's okay, I can fix it myself later. 🙂

@superus8r what fix your are referring to? Could you elaborate please.

**Comment by superus8r** [comment]
@prempratapsingh 

> The increase in code coverage reports is primarily due to the refactoring I did last week for Views, Models, ViewModel and other related layers. Please see this code commit for reference

I understand and respect that 🙂 

This discussion is:
- about: `tests covering themselves` 
- not about: `increase in code coverage` ❌

Please do not mix these.

**Comment by prempratapsingh** [comment]
> @prempratapsingh
> 
> > The increase in code coverage reports is primarily due to the refactoring I did last week for Views, Models, ViewModel and other related layers. Please see this code commit for reference
> 
> I understand and respect that 🙂
> 
> This discussion is:
> 
> * about: `tests covering themselves`
> * not about: `increase in code coverage` ❌
> 
> Please do not mix these.

@superus8r I am confused about why there is point/concern about `tests covering themselves` now?

Please see Xcode generated code coverage report (for `feature/buyer-protection` branch) in the in the screenshot below. This coverage report is purely by Xcode, there is no CodeCov or CircleCI envolved, still the coverage shows 93.6% :)

Screenshot 2023-04-03 at 3 22 45 PM

Anyways, if you feel like some CircleCI end of change is required for better test coverage, please go for it.

**Comment by superus8r** [comment]
@prempratapsingh 
Thanks I will do that

File: Sources/Trustylib/Utilities/GradeCalculator.swift
Comment (prempratapsingh) [reply Sources/Trustylib/Utilities/GradeCalculator.swift:1]: The same rating to grade conversion logic has to be used for shop grade and product grade widgets, this is why this I added this GradeCalculator for reusability and improved unit tests.

**Comment by hrayatnia** [comment]
In spirits of open source and contribution, two of the tests get failed with XCode 14.2 by just running tests and also by using Fastlane.
Iphone 11 simulator, Iphone 14 Pro simulator.

Test names: 

`testBuyerProtectionDetailsReturnsCorrectFormattedStringForProtectionAmount` on both devices, regardless of the approach

`testNetworkDataServiceAddsCorrectURLParameters` acted randomly but failed a couple of times.

Screenshot 2023-04-03 at 12 01 38

PNG image

PNG image



Also, to contribute to the discussion, the XCode test coverage will also be 91.8% if you turn on the coverage target.

Screenshot 2023-04-03 at 12 24 31

Screenshot 2023-04-03 at 12 24 42

**Comment by superus8r** [comment]
@prempratapsingh fyi
